Sources/Crypto/BCrypt/BCryptDigest.swift

```diff
@@ -1,37 +1,118 @@
-        let config = try BCryptConfig(
-            version: .two(.y),
-            cost: cost,
-            salt: salt?.convertToData() ?? CryptoRandom().generateData(count: 16)
+    public func hash(_ plaintextData: LosslessDataConvertible, cost: UInt = 12, salt saltData: LosslessDataConvertible) throws -> String {
+
+        let salt = String.convertFromData(saltData.convertToData())
+        guard isSaltValid(salt) else {
+            throw CryptoError(identifier: "invalidSalt", reason: "Provided salt has the incorrect format")
+        }
+
+        let originalAlgorithm: Algorithm
+        if salt.count == Algorithm.saltCount {
+            // user provided salt
+            originalAlgorithm = ._2b
+        } else {
+            // full salt, not user provided
+            let revisionString = String(salt.prefix(4))
+            if let parsedRevision = Algorithm(rawValue: revisionString) {
+                originalAlgorithm = parsedRevision
+            } else {
+                throw CryptoError(identifier: "invalidSalt", reason: "Provided salt has the incorrect format")
+            }
+        }
+
+        // OpenBSD doesn't support 2y revision.
+        let normalizedSalt: String
+        if originalAlgorithm == Algorithm._2y {
+            // Replace with 2b.
+            normalizedSalt = Algorithm._2b.rawValue + salt.dropFirst(originalAlgorithm.revisionCount)
+        } else {
+            normalizedSalt = salt
+        }
+
+        let plaintext = String.convertFromData(plaintextData.convertToData())
+        let hashedBytes = UnsafeMutablePointer<Int8>.allocate(capacity: 128)
+        defer { hashedBytes.deallocate() }
+        let hashingResult = bcrypt_hashpass(
+            plaintext,
+            normalizedSalt,
+            hashedBytes,
+            128
         )
-        let digest = try BCryptAlgorithm(config: config).digest(message: plaintext.convertToData())
-        let serializer = BCryptSerializer(config: config, digest: digest)
-        return serializer.serialize()
+
+        if hashingResult != 0 {
+            throw CryptoError(identifier: "unableToComputeHash", reason: "Unable to compute BCrypt hash")
+        } else {
+            return originalAlgorithm.rawValue + String(cString: hashedBytes).dropFirst(originalAlgorithm.revisionCount)
+        }
+    }
+
+    /// Creates a BCrypt digest for the supplied data. `salt` will be generated.
+    ///
+    ///     try BCrypt.hash("vapor", cost: 4)
+    ///
+    /// - parameters:
+    ///     - plaintext: Plaintext data to hash.
+    ///     - cost: Desired complexity. Larger `cost` values take longer to hash and verify.
+    /// - throws: `CryptoError` if hashing fails or if data conversion fails.
+    /// - returns: BCrypt hash for the supplied plaintext data.
+    public func hash(_ plaintext: LosslessDataConvertible, cost: UInt = 12) throws -> String {
```

Comment:
I think this could be represented by making salt optional in the previous method. That would also prevent another breaking change that isn't really necessary 👍 (unless this was done for a reason).

Reply:
I just didn't want to use an extra optional. But it's really up to you (extra public function vs optional parameter)

Reply:
This change would break:

```swift
let hash = try BCrypt.hash("foo", salt: nil)
```

Which works currently. There should be 0 breaking changes in a minor release. The Data -> String thing is an exception because that is a bug really.